pkg/ansible/runner.go

```diff
@@ -0,0 +1,201 @@
+	if err != nil {
+		return err
+	}
+
+	jobName := "openshift-ansible-test-job"
+	env := []kapi.EnvVar{
+		{
+			Name:  "INVENTORY_FILE",
+			Value: "/ansible/inventory/hosts",
+		},
+		{
+			Name:  "PLAYBOOK_FILE",
+			Value: playbook,
+		},
+		{
+			Name:  "ANSIBLE_HOST_KEY_CHECKING",
+			Value: "False",
+		},
+		{
+			Name:  "OPTS",
+			Value: "-vvv --private-key=/ansible/ssh/privatekey.pem -e @/ansible/inventory/vars",
+		},
+		{
+			Name: "AWS_ACCESS_KEY_ID",
+			ValueFrom: &kapi.EnvVarSource{
+				SecretKeyRef: &kapi.SecretKeySelector{
+					LocalObjectReference: kapi.LocalObjectReference{Name: awsCredentialsSecret},
+					Key:                  "aws_access_key_id",
+				},
+			},
+		},
+		{
+			Name: "AWS_SECRET_ACCESS_KEY",
+			ValueFrom: &kapi.EnvVarSource{
+				SecretKeyRef: &kapi.SecretKeySelector{
+					LocalObjectReference: kapi.LocalObjectReference{Name: awsCredentialsSecret},
+					Key:                  "aws_secret_access_key",
+				},
+			},
+		},
+	}
+	runAsUser := int64(0)
+	sshKeyFileMode := int32(0600)
+	playbookPath := path.Join(openshiftAnsibleContainerDir, playbook)
+	podSpec := kapi.PodSpec{
+		DNSPolicy:          kapi.DNSClusterFirst,
+		RestartPolicy:      kapi.RestartPolicyNever,
+		ServiceAccountName: openshiftAnsibleServiceAccount,
+		//HostNetwork:        true,
+
+		Containers: []kapi.Container{
+			{
+				Name:  jobName,
+				Image: r.Image,
+				Env:   env,
+				SecurityContext: &kapi.SecurityContext{
+					RunAsUser: &runAsUser,
+				},
+				VolumeMounts: []kapi.VolumeMount{
+					{
+						Name:      "inventory",
+						MountPath: "/ansible/inventory/",
+					},
+					{
+						Name:      "sshkey",
+						MountPath: "/ansible/ssh/",
+					},
+				},
+				//Command: []string{"sleep", "1000000"},
+
+				// TODO: drop this once https://github.com/openshift/openshift-ansible/pull/6320 merges, the default run script should then work, minus the vars file :(
+				Command: []string{"ansible-playbook", "-i", "/ansible/inventory/hosts", "--private-key", "/ansible/ssh/privatekey.pem", playbookPath, "-e", "@/ansible/inventory/vars"},
+			},
+		},
+		Volumes: []kapi.Volume{
+			{
+				// Mounts both our inventory and vars file.
+				Name: "inventory",
+				VolumeSource: kapi.VolumeSource{
+					ConfigMap: &kapi.ConfigMapVolumeSource{
+						LocalObjectReference: kapi.LocalObjectReference{
+							Name: inventoryConfigMap,
+						},
+					},
+				},
+			},
+			{
+				Name: "sshkey",
+				VolumeSource: kapi.VolumeSource{
+					Secret: &kapi.SecretVolumeSource{
+						SecretName: sshPrivateKeySecret,
+						Items: []kapi.KeyToPath{
+							{
+								Key:  "ssh-privatekey",
+								Path: "privatekey.pem",
+								Mode: &sshKeyFileMode,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	completions := int32(1)
+	deadline := int64(60 * 60) // one hour for now
+
+	meta := metav1.ObjectMeta{
+		GenerateName: jobNamePrefix,
+		Namespace:    operatorNamespace,
+	}
+
+	job := &kbatch.Job{
```

Comment:
I'm wondering whether a job is the most appropriate resource here. The job controller will re-run the job blindly until it either succeeds or the completion deadline expires. I'm wondering if we'd want to use our own controller to manage the job (and implement things like exponential backoff and fixed number of retries, etc). If that's the case, then it's better off as a run-once pod.

Reply:
With the understanding that I have no experience with jobs, they seem to me to be the appropriate thing to use here. My understanding is that the job controller already does exponential backoff of failed jobs. I don't see much value in having a fixed number of retries over a deadline for retries.

Reply:
Fair point

Reply:
Yeah we did plan to have this thing re-try until completion so the Job felt like a good fit, at least for now. I'd pitch we stick with Job until it starts to fail us.